### PR TITLE
Rewrite string info

### PR DESCRIPTION
**Pull request by eeeebbbbrrrr**
Rejigger our `StringInfo` wrapper to use a PgBox and honor the various "WhoAllocated" traits.
    
I think it's pretty important for the compiler to have an understanding of who is going to drop a StringInfo, as it's used in the "inoutfuncs" portions of PostgresType, and that's going to get an overhaul soon.

This wants #901 to be merged first.

### CODE REVIEW

**Comment by eeeebbbbrrrr** [comment]
I've rebased this PR.  @thomcc, could I get a review?  I think this is good cleanup for `StringInfo`

File: pgx/src/stringinfo.rs [hunk 1]
```diff
@@ -131,144 +138,166 @@ impl StringInfo {
...
+                bytes.len().try_into().expect("len of bytes doesn't fit in an i32"),
             )
         }
     }
 
     /// Push the bytes behind a raw pointer of a given length onto the end
     #[inline]
-    pub fn push_raw(&mut self, ptr: void_mut_ptr, len: usize) {
-        // safe:  self.sid will never be null
+    pub unsafe fn push_raw(&mut self, ptr: void_mut_ptr, len: i32) {
```
[the remark this answers is not in the document]
Comment (workingjubilee) [reply pgx/src/stringinfo.rs:161]: Q: should signature be `*const u8` to signify we cast to bytes?

File: pgx/src/stringinfo.rs [hunk 2]
```diff
@@ -65,51 +59,64 @@ impl std::io::Write for StringInfo {
...
 
     /// Construct a new `StringInfo`, allocated by Postgres in `CurrentMemoryContext`, ensuring it
     /// has a capacity of the specified `len`.
     ///
     /// Note that Postgres can only represent up to 1 gigabyte of data in a `StringInfo`
-    ///
-    /// Unless `.into_pg()` or `.into_char_ptr()` are called, memory management of
-    /// this `StringInfo` follow Rust's drop semantics.
-    pub fn with_capacity(len: usize) -> Self {
+    pub fn with_capacity(len: i32) -> Self {
```
Comment (workingjubilee) [inline pgx/src/stringinfo.rs:89]: Hm, why the change to the `i32` type?

File: pgx/src/stringinfo.rs [hunk 3]
```diff
@@ -131,144 +138,166 @@ impl StringInfo {
...
+    /// A `&str` representation.
+    ///
+    /// # Errors
+    ///
+    /// If the contained bytes aren't valid UTF8, a [UTF8Error] is returned.  Postgres StringInfo
+    /// is allowed to contain non-UTF8 byte sequences, so this is a real possibility.
+    #[inline]
+    pub fn as_str(&self) -> Result<&str, Utf8Error> {
+        std::str::from_utf8(self.as_bytes())
+    }
```
Comment (workingjubilee) [inline pgx/src/stringinfo.rs:196]: Nice, was actually about to ask if this could exist!

File: pgx/src/stringinfo.rs (hunk 2 above)
Comment (eeeebbbbrrrr) [reply pgx/src/stringinfo.rs:89]: I argued with myself over this.  Postgres' `StringInfo` length is an `i32`, so I thought rather than us deal with an `usize` that overflows an i32 and have to return a Result, just let the user sort it out ahead of time.

Thoughts?

File: pgx/src/stringinfo.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (eeeebbbbrrrr) [reply pgx/src/stringinfo.rs:161]: well, I suppose it should be `*const std::os::raw::c_char`, as again, that's what Postgres expects

File: pgx/src/stringinfo.rs [hunk 4]
```diff
@@ -65,51 +59,64 @@ impl std::io::Write for StringInfo {
...
-impl ToString for StringInfo {
+impl<AllocatedBy: WhoAllocated> ToString for StringInfo<AllocatedBy> {
+    /// Convert this [`StringInfo`] into a Rust string.  This uses [`String::from_utf8_lossy`] as
+    /// it's fine for a Postgres [`StringInfo`] to contain null bytes and also not even be proper
+    /// UTF8.
     fn to_string(&self) -> String {
-        unsafe { std::str::from_utf8_unchecked(self.as_bytes()).to_owned() }
+        String::from_utf8_lossy(self.as_bytes()).to_string()
     }
 }
```
[the remark this answers is not in the document]
Comment (eeeebbbbrrrr) [reply pgx/src/stringinfo.rs]: Changed.  I didn't realize that implementing `Display` got you `ToString` for free.  TIL.

File: pgx/src/stringinfo.rs (hunk 2 above)
Comment (workingjubilee) [reply pgx/src/stringinfo.rs:90]: We discussed this on Discord and my sentiment is broadly "`usize` is fine and indicates what the input is "for" to Rust programmers (morally it's actually `pg_size_t`), and Postgres is running so many checks that it will throw a fit if we throw a bad input into Postgres, don't worry about it". `with_capacity` is really more of an "optimization hint" to Postgres, as Postgres will, before pushing anything on StringInfo, always calculate `needed` before pushing data into it.

**Review by workingjubilee — APPROVED** [review]
Despite my stated inclination, I think the type thing can ultimately go either way so I am good with this now.
